src/ppx/grfx/grfx_sync.cpp

```diff
@@ -32,5 +37,60 @@ Result Fence::WaitAndReset(uint64_t timeout)
     return ppx::SUCCESS;
 }
 
+// -------------------------------------------------------------------------------------------------
+// Semaphore
+// -------------------------------------------------------------------------------------------------
+Result Semaphore::Wait(uint64_t value, uint64_t timeout) const
+{
+    if (this->GetSemaphoreType() != grfx::SEMAPHORE_TYPE_TIMELINE) {
+        PPX_ASSERT_MSG(false, REQUIRES_TIMELINE_MSG);
+        return ppx::ERROR_GRFX_INVALID_SEMAPHORE_TYPE;
+    }
+
+    auto ppxres = this->TimelineWait(value, timeout);
+    if (Failed(ppxres)) {
+        return ppxres;
+    }
+
+    return ppx::SUCCESS;
+}
+
+Result Semaphore::Signal(uint64_t value, bool forceMonotonicValue) const
+{
+    if (this->GetSemaphoreType() != grfx::SEMAPHORE_TYPE_TIMELINE) {
+        PPX_ASSERT_MSG(false, REQUIRES_TIMELINE_MSG);
+        return ppx::ERROR_GRFX_INVALID_SEMAPHORE_TYPE;
+    }
+
+    // Synchronize access to API semaphore object
+    std::lock_guard<std::mutex> lock(mTimelineMutex);
+
+    if (forceMonotonicValue) {
```

Comment:
Since sending an lower value can cause locks or failures, shall we assert if such behavior is detected? Or is there some obscure valid pattern that requires this?

Reply:
It's probably more accurately described as naivety prevention. There's some Vulkan literature out there that mention more than a few times that you can use timeline semaphore to signals on the CPU on different threads. What I didn't see mention was what happens if the threads signal out of order? Is the application programmer suppose to put in some kind of measure where the threads need to signal in order? What would that look like? Seems liek that might be repetitive work on the application side. This flag just makes the out of order signaling from different threads case easier to deal with.